keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+
+For a simple trial, I create a Pod directly where two containers share the same subset
+of the GPU:
+```
+apiVersion: gpu.acme.com/v1
+kind: GPURequirements
+metadata:
+  name: device-consumer-gpu-parameters
+memory: "2Gi"
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  resourceClaims:
+  - name: "gpu" # this name gets referenced below under "claims"
+    template:
+      resourceClassName: "acme-gpu"
+      parameters:
+        apiVersion: gpu.acme.com/v1
+        kind: GPURequirements
+        name: device-consumer-gpu-parameters
+  containers:
+  - name: workload
+    image: my-app
+    command: ["/bin/program"]
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "128Mi"
+        cpu: "500m"
+      claims:
+        - "gpu"
+  - name: monitor
+    image: my-app
+    command: ["/bin/other-program"]
+    resources:
+      requests:
+        memory: "32Mi"
+        cpu: "25m"
+      limits:
+        memory: "64Mi"
+        cpu: "50m"
+      claims:
+      - "gpu"
+```
+
+This request triggers resource allocation on a node that has a GPU device with
+2Gi of memory available and then the Pod runs on that node. The remaining
+capacity of the GPU can be used by other pods. The lifecycle of the resource
+allocation is tied to the lifecycle of the Pod.
+
+In production, a similar PodTemplateSpec in a Deployment will be used.
+
+#### Combined setup of different hardware functions
+
+As a 5G telco operator, I want to use the FPGA IP block, signal processor and
+network interfaces provided by the [Intel FPGA
+N3000](https://www.intel.com/content/www/us/en/products/details/fpga/platforms/pac/n3000.html)
+card in a Kubernetes edge cluster. Intel provides a single resource driver that
+has parameters for setting up all of these hardware functions together as
+needed for a data flow pipeline.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Scheduling is likely to be slower when many Pods request the new
+resource types, both because scheduling such a Pod involves more
+round-trips through the API server for ResourceClaimStatus updates and
+because scheduling one Pod may affect other Pods in ways that cannot
+be anticipated by the kube-scheduler. When many Pods compete for
+limited resources, multiple attempts may be needed before a suitable
+node is found.
+
+The hardware that is expected to need this more flexible allocation
+approach is going to be used by only a small subset of the pods in the
+cluster and those pods are likely to run for extended periods of time,
+so this is not a major concern.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
```

Comment:
Nothing here?  The discussion above seems germane.  scheduler throughput seems worth documenting.  How abouot security of these plugins - putting them inline with scheduling seems like a good way to break clusters or to steer sensitive work onto compromised nodes.  E.g. I compromiose a node which happens to run a driver plugin.  Can I use that to ensure the workloads I care about land on this node, where I already own the OS?

Have we had someone from sig-security think about this?

Reply:
I have pushed an update with text for this section. Given how problematic write access for kubelet to ResourceClaimStatus is I am now almost leaning towards removing that from the design - it'll just take a small nudge in that direction.

Reply:
+1 on @mrunalp questions listed above. I am ok this to move forward with those open questions for alpha. But we need to address them for beta for sure.